### PR TITLE
Metro mode selector: coreutils

### PR DESCRIPTION
**Pull request by daniel-heppner-ibigroup**
The core utils changes are mostly focused on generating OTP GraphQL Queries to go along with the new mode selector and associated changes. We also have code to generate combinations, so they no longer need to be specified by the consumer.

### CODE REVIEW

**Comment by daniel-heppner-ibigroup** [comment]
Blocked until #551 is merged.

File: packages/core-utils/src/planQuery.graphql [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	$intermediatePlaces: [InputCoordinates],
+	$preferred: InputPreferred,
+	$unpreferred: InputUnpreferred,
+	$banned: InputBanned,
+) {
+	plan(
+		fromPlace: $fromPlace
+		toPlace: $toPlace
+		transportModes: $modes
+		locale: "en",
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/planQuery.graphql:21]: Add a comment regarding what setting the locale does.

File: packages/core-utils/src/query-gen.ts [hunk 2]
```diff
@@ -0,0 +1,205 @@
...
+      return [...prev, cur.addTransportMode];
+    }
+    if (cur.type === "DROPDOWN") {
+      const transportMode = cur.options.find(o => o.value === cur.value)
+        .addTransportMode;
+      if (transportMode) {
+        return [
+          ...prev,
+          cur.options.find(o => o.value === cur.value).addTransportMode
+        ];
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query-gen.ts:51]:
```suggestion
        return [...prev, transportMode];
```

File: packages/core-utils/src/query-gen.ts [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+  return (
+    // eslint-disable-next-line no-bitwise
+    new Array(1 << array.length)
+      .fill(null)
+      // eslint-disable-next-line no-bitwise
+      .map((e1, i) => array.filter((e2, j) => i & (1 << j)))
+  );
+}
+
+export const SIMPLIFICATIONS = {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query-gen.ts:76]: Add a comment that this variable maps each mode into a broader travel mode.

File: packages/core-utils/src/query-gen.ts [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+      return false;
+    }
+
+    if (
+      combo.reduce((prev, cur) => {
+        if (queryTransitSubmodes.includes(cur.mode)) {
+          return prev - 1;
+        }
+        return prev;
+      }, queryTransitSubmodes.length) !== 0
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query-gen.ts:138]: Extract a variable, so it's easier to read what this condition is about.

File: packages/core-utils/src/query-gen.ts [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+  return !!VALID_COMBOS.find(
+    vc =>
+      simplifiedModes.every(m => vc.includes(m)) &&
+      vc.every(m => simplifiedModes.includes(m))
+  );
+}
+
+/**
+ * Generates list of queries for OTP to get a comprehensive
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query-gen.ts:156]:
```suggestion
 * Generates a list of queries for OTP to get a comprehensive
 * ```

File: packages/core-utils/src/query-gen.ts [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+    .filter(mode => TRANSIT_SUBMODES.includes(mode.mode))
+    .map(mode => mode.mode);
+
+  return combinations(completeModeList)
+    .filter(combo => isCombinationValid(combo, queryTransitSubmodes))
+    .map(combo => ({ ...params, modes: combo }));
+}
+
+export function generateOtp2Query(params: OTPQueryParams): GraphQLQuery {
+  const { from, modeSettings, to } = params;
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query-gen.ts:178]: You might be able to move this line to the function declaration.

File: packages/core-utils/src/query-gen.ts [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+    wheelchair
+  } = modeSettingValues;
+
+  return {
+    query: print(PlanQuery),
+    variables: {
+      bikeReluctance,
+      carReluctance,
+      fromPlace: `${from.name}::${from.lat},${from.lon}}`,
+      modes: params.modes,
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/src/query-gen.ts:199]: Declare `modes` within line 178.

File: packages/core-utils/package.json [hunk 8]
```diff
@@ -12,6 +12,7 @@
   "license": "MIT",
   "private": false,
   "dependencies": {
```
Comment (binh-dam-ibigroup) [inline packages/core-utils/package.json:14]: You will have to update the version of `@opentripplanner/types` under `devDependencies`.

**Review by binh-dam-ibigroup — APPROVED** [review]
Some minor cleanup and a conflict to fix, but otherwise looks good.

File: packages/core-utils/src/__tests__/query-params.ts [hunk 9]
```diff
@@ -0,0 +1,279 @@
...
+    value: true,
+    addTransportMode: mode
+  };
+
+  const dropdownModeSetting: ModeSetting = {
+    type: "DROPDOWN",
+    label: "test",
+    applicableMode: testTransportMode.mode,
+    key: "test",
+    options: [{ text: "testop", value: "1", addTransportMode: mode }],
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/core-utils/src/__tests__/query-params.ts:78]: It was test option, but I made it more clear.

**Review by AdrianaCeric — APPROVED** [review]
I have limited knowledge of the new mode selector and OTP flex compatible modes so my review isn't too in-depth here, but code makes sense!

File: packages/core-utils/src/__tests__/query-params.ts [hunk 10]
```diff
@@ -0,0 +1,279 @@
...
+    value: true
+  };
+
+  const dropdownModeSetting: ModeSetting = {
+    applicableMode: testTransportMode.mode,
+    key: "test",
+    label: "test",
+    options: [{ text: "testOption", value: "1", addTransportMode: mode }],
+    type: "DROPDOWN",
+    value: "1"
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/core-utils/src/__tests__/query-params.ts:79]: Thanks for the suggestion, I've added it.

File: packages/core-utils/src/__tests__/query-params.ts [hunk 11]
```diff
@@ -0,0 +1,279 @@
...
+      [["TRANSIT"], ["FLEX", "TRANSIT"], ["FLEX", "WALK"], ["WALK"]]
+    );
+    expectModes(
+      ["FLEX", "SCOOTER_RENT", "TRANSIT", "WALK"],
+      [
+        ["TRANSIT"],
+        ["FLEX", "TRANSIT"],
+        ["WALK"],
+        ["FLEX", "WALK"],
+        ["FLEX", "SCOOTER_RENT", "WALK"], // Is this sensible?
```
[the remark this answers is not in the document]
Comment (daniel-heppner-ibigroup) [reply packages/core-utils/src/__tests__/query-params.ts:190]: Miles wrote that, but I think the question is whether using scooter_rent and walk can be used together. (the answer is yes I think)

**Pull request merged**